File: source/409.最长回文串.py
#
# @lc app=leetcode.cn id=409 lang=python3
#
# [409] 最长回文串
#

# @lc code=start
class Solution:
    def longestPalindrome(self, s: str) -> int:
        count = [0]*128
        for c in s:
            count[ord(c)] += 1
        length = 0

        for i in count:
            length += i - (i&1)

        return length+1 if length < len(s) else length


# @lc code=end

File: source/409.最长回文串.cpp
/*
 * @lc app=leetcode.cn id=409 lang=cpp
 *
 * [409] 最长回文串
 */
#include "cpp_includes.h"
// @lc code=start
class Solution {
   public:
    int longestPalindrome(string s) {
        vector<int> count(128, 0);
        for (auto c : s) {
            count[c]++;
        }

        int len = 0;
        for (auto cnt : count) {
            len += cnt - (cnt & 1);
        }

        return len < s.length() ? len + 1 : len;
    }
};
// @lc code=end
